### PR TITLE
JAV-30 Auto Configration code for spring-boot-starter-discovery

### PR DESCRIPTION
**Pull request by asifhuawei**
Auto Service Registration for spring-boot-starter-discovery

### CODE REVIEW

File: java-chassis-dependencies/pom.xml [hunk 1]
```diff
@@ -1003,6 +1009,16 @@
 			</dependency>
 			<dependency>
 				<groupId>io.servicecomb</groupId>
+				<artifactId>foundation-crypt</artifactId>
+				<version>0.1.0-m2-SNAPSHOT</version>
```
Comment (seanyinx) [inline java-chassis-dependencies/pom.xml]: the conflict resolving with master is not wring. these dependencies were missing and removed on master.

File: java-chassis-dependencies/pom.xml [hunk 2]
```diff
@@ -1059,6 +1075,11 @@
 			</dependency>
 			<dependency>
 				<groupId>io.servicecomb</groupId>
+				<artifactId>transport-grpc</artifactId>
```
Comment (seanyinx) [inline java-chassis-dependencies/pom.xml]: same as above

File: spring-boot-starter/spring-boot-starter-discovery/pom.xml [hunk 3]
```diff
@@ -116,6 +96,10 @@
 		</dependency>
 		<dependency>
 			<groupId>io.servicecomb</groupId>
+			<artifactId>transport-grpc</artifactId>
```
Comment (seanyinx) [inline spring-boot-starter/spring-boot-starter-discovery/pom.xml]: grpc module was removed already

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistry.java [hunk 4]
```diff
@@ -20,21 +20,23 @@
...
-		try {
-			this.client.init();
-		} catch (Exception e) {
-			// TODO Auto-generated catch block
-			e.printStackTrace();
+	public CseServiceRegistry() {
+		if (null == this.client) {
+			this.client = RegistryClientFactory.getRegistryClient();
+			try {
+				this.client.init();
```
Comment (seanyinx) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistry.java]: please rebase onto branch JAV-30_discovery_client. this method shall not throw any exception.

File: java-chassis-dependencies/pom.xml [hunk 5]
```diff
@@ -116,6 +117,11 @@
 				<version>${spring-cloud-security.version}</version>
 			</dependency>
 			<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-netflix-core</artifactId>
+			<version>${spring-cloud-netflix.version}</version>
```
Comment (seanyinx) [inline java-chassis-dependencies/pom.xml]: indentation is not right

File: demo/demo-spring-boot-discovery/demo-spring-boot-discovery-client/src/test/java/io/servicecomb/demo/discovery/client/DiscoveryClientIT.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import static org.junit.Assert.assertThat;
+
+public class DiscoveryClientIT {
+
+	@Before
+	public void setUp() {
+		TestMgr.errors().clear();
+	}
+
+	@Test
```
Comment (seanyinx) [inline demo/demo-spring-boot-discovery/demo-spring-boot-discovery-client/src/test/java/io/servicecomb/demo/discovery/client/DiscoveryClientIT.java]: this test doesn't have much meaning except ensured DiscoveryClient can be started up.

Nothing related to how services discovery with spring cloud integration and load balancing was tested.

We have to be sure we are able to forward requests to corresponding services from zuul with service center

File: java-chassis-dependencies/pom.xml (hunk 1 above)
Comment (asifhuawei) [inline java-chassis-dependencies/pom.xml]: okay, will fix it

File: java-chassis-dependencies/pom.xml (hunk 2 above)
Comment (asifhuawei) [inline java-chassis-dependencies/pom.xml]: okay, will remove this

File: spring-boot-starter/spring-boot-starter-discovery/pom.xml (hunk 3 above)
Comment (asifhuawei) [inline spring-boot-starter/spring-boot-starter-discovery/pom.xml]: okay will remove this

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistry.java (hunk 4 above)
Comment (asifhuawei) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistry.java]: okay will remove try_catch block

File: java-chassis-dependencies/pom.xml (hunk 5 above)
Comment (asifhuawei) [inline java-chassis-dependencies/pom.xml]: okay

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistration.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package io.servicecomb.springboot.starter.serviceregistry;
```
Comment (WillemJiang) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistration.java]: Please add the license header here.

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistrationAutoConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package io.servicecomb.springboot.starter.serviceregistry;
```
Comment (WillemJiang) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistrationAutoConfiguration.java]: License header.

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistryAutoConfiguration.java [hunk 9]
```diff
@@ -23,16 +23,14 @@
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
-import io.servicecomb.serviceregistry.client.ServiceRegistryClient;
-
 @Configuration
 @ConditionalOnProperty(value = "spring.cloud.service-registry.enabled", matchIfMissing = true)
 @AutoConfigureBefore(ServiceRegistryAutoConfiguration.class)
 public class CseServiceRegistryAutoConfiguration {
```
Comment (WillemJiang) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistryAutoConfiguration.java]: We should avoid using CSE any more.

File: demo/demo-spring-boot-discovery/demo-spring-boot-discovery-client/src/test/java/io/servicecomb/demo/discovery/client/DiscoveryClientIT.java (hunk 6 above)
Comment (asifhuawei) [inline demo/demo-spring-boot-discovery/demo-spring-boot-discovery-client/src/test/java/io/servicecomb/demo/discovery/client/DiscoveryClientIT.java]: Written demo/IT code for:
1. To test the discovery of particular micro-service
2. To test the Zuul Proxy gateway feature to redirect the request to proper micro-service.

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistration.java (hunk 7 above)
Comment (asifhuawei) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistration.java]: added

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistrationAutoConfiguration.java (hunk 8 above)
Comment (asifhuawei) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseAutoServiceRegistrationAutoConfiguration.java]: added

File: spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistryAutoConfiguration.java (hunk 9 above)
Comment (asifhuawei) [inline spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/serviceregistry/CseServiceRegistryAutoConfiguration.java]: okay once this PR is merged I will re-factor the names and raise another MR immediately, right now the PR code size is big, re-merging/re-basing is adding extra effort.

**Comment by seanyinx** [comment]
could you add this `<module>demo-spring-boot-discovery</module>` to demo-parent to enable it for testing? I can't get it pass on my laptop.

**Comment by asifhuawei** [comment]
Added the "spring-boot-starter-discovery" and "demo-spring-boot-starter-discovery" in parent pom

**Pull request merged**